Documentation/man/man3/GameObject.3
.TH "GameObject" 3 "Version 0.0.0.2" "My Project" \" -*- nroff -*-
.ad l
.nh
.SH NAME
GameObject
.SH SYNOPSIS
.br
.PP
.PP
\fR#include <GAMEOBJECT\&.h>\fP
.PP
Inherits \fBCORE::Object\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBGameObject\fP ()"
.br
.ti -1c
.RI "\fBGameObject\fP (int null)"
.br
.ti -1c
.RI "\fBGameObject\fP (const \fBGameObject\fP &other)"
.br
.ti -1c
.RI "\fBGameObject\fP & \fBoperator=\fP (const \fBGameObject\fP &other)"
.br
.ti -1c
.RI "std::shared_ptr< \fBCORE::Behaviour\fP > \fBAddComponent\fP (std::shared_ptr< \fBCORE::Behaviour\fP > component, std::string \fBname\fP)"
.br
.ti -1c
.RI "template<class _Ty , class\&.\&.\&. _Types> _Ty * \fBAddComponent\fP (_Ty arg)"
.br
.ti -1c
.RI "template<class _Ty , class\&.\&.\&. _Types> void \fBAddComponent\fP (_Ty arg, void *component)"
.br
.ti -1c
.RI "template<class _Ty , class\&.\&.\&. _Types> _Ty * \fBGetComponent\fP (_Ty arg)"
.br
.ti -1c
.RI "template<class _Ty , class\&.\&.\&. _Types> _Ty * \fBGetComponentDynamic\fP (_Ty arg)"
.br
.in -1c

Public Member Functions inherited from \fBCORE::Object\fP
.in +1c
.ti -1c
.RI "\fBObject\fP ()"
.br
.ti -1c
.RI "virtual \fB~Object\fP ()"
.br
.ti -1c
.RI "void \fBDestroyObj\fP (\fBObject\fP *obj)"
.br
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static std::shared_ptr< \fBGameObject\fP > \fBCreate\fP (\fBDATATYPES::TS_P_Vector3\fP pos, void *mesh, int materialID)"
.br
.ti -1c
.RI "static shared_ptr< \fBGameObject\fP > \fBCreate\fP (\fBDATATYPES::TS_P_Vector3\fP pos, \fBQuaternion\fP rot, std::shared_ptr< \fBCOMPONENTS::_Mesh\fP > mesh, int materialID)"
.br
.RI "Creates new \fBGameObject\fP\&. "
.ti -1c
.RI "static \fBGameObject\fP \fBCreateEmpty\fP (\fBDATATYPES::TS_P_Vector3\fP pos)"
.br
.ti -1c
.RI "static \fBGameObject\fP \fBCreateCamera\fP (\fBDATATYPES::TS_P_Vector3\fP pos, \fBQuaternion\fP rot)"
.br
.in -1c
.SS "Public Attributes"

.in +1c
.ti -1c
.RI "bool \fBEnabled\fP = true"
.br
.ti -1c
.RI "std::shared_ptr< \fBCOMPONENTS::_Mesh\fP > \fBMESH\fP"
.br
.ti -1c
.RI "std::shared_ptr< \fBTransform\fP > \fBTrAnSfOrM\fP"
.br
.ti -1c
.RI "std::shared_ptr< \fBCORE::Behaviour\fP > \fBMesH\fP"
.br
.ti -1c
.RI "\fBTag\fP \fBtag\fP"
.br
.ti -1c
.RI "\fBLayer\fP \fBlayer\fP"
.br
.ti -1c
.RI "void * \fBmsh\fP"
.br
.ti -1c
.RI "const bool \fBisNull\fP"
.br
.ti -1c
.RI "bool \fBisRemoved\fP"
.br
.ti -1c
.RI "std::string \fBname\fP"
.br
.ti -1c
.RI "std::vector< void * > \fBComponents\fP"
.br
.ti -1c
.RI "std::vector< std::shared_ptr< \fBCORE::Behaviour\fP > > \fBbehaviours\fP"
.br
.ti -1c
.RI "std::vector< UUID > \fBuuids\fP"
.br
.in -1c

Public Attributes inherited from \fBCORE::Object\fP
.in +1c
.ti -1c
.RI "UUID \fBuuid\fP"
.br
.in -1c
.SH "Detailed Description"
.PP 
Definition at line \fB23\fP of file \fBGAMEOBJECT\&.h\fP\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "GameObject::GameObject ()\fR [inline]\fP"

.PP
Definition at line \fB39\fP of file \fBGAMEOBJECT\&.h\fP\&.
.nf
39 :isNull(false), isRemoved(false) {};
.PP
.fi

.SS "GameObject::GameObject (int null)\fR [inline]\fP"

.PP
Definition at line \fB40\fP of file \fBGAMEOBJECT\&.h\fP\&.
.nf
40 :isNull(true), isRemoved(true) {};
.PP
.fi

.SS "GameObject::GameObject (const \fBGameObject\fP & other)\fR [inline]\fP"

.PP
Definition at line \fB41\fP of file \fBGAMEOBJECT\&.h\fP\&.
.nf
42             : Enabled(other\&.Enabled),
43             MESH(other\&.MESH),
44             TrAnSfOrM(other\&.TrAnSfOrM),
45             MesH(other\&.MesH),
46             msh(other\&.msh),
47             isNull(other\&.isNull),
48             name(other\&.name),
49             Components(other\&.Components),
50             behaviours(other\&.behaviours),
51             uuids(other\&.uuids),
52             isRemoved(false)
53         {}
.PP
.fi

.SH "Member Function Documentation"
.PP 
.SS "template<class _Ty , class\&.\&.\&. _Types> _Ty * GameObject::AddComponent (_Ty arg)\fR [inline]\fP"

.PP
Definition at line \fB81\fP of file \fBGAMEOBJECT\&.h\fP\&.
.nf
81                                    {
82             TracyCZoneN(ctxx, "Adding Component", true);
83 
84             if (arg\&.UUID == GUID_NULL) {
85                 UUID uuid;
86                 UuidCreate(&uuid);
87 
88 
89                 arg\&.UUID = uuid;
90                 arg\&.uuID = arg\&.UUID;
91             }
92             uuids\&.push_back(arg\&.UUID);
93 
94             auto aaa = arg;
95             aaa\&.gameobject = this;
96             //aaa\&.transf = behaviours[1];
97             aaa\&.TRANSFORM = TrAnSfOrM;
98             //using name = decltype(_Ty);
99                 aaa\&.compname = typeid(_Ty)\&.name();
100             //aa\->gameobject = this;
101             // 
102             std::shared_ptr<CORE::Behaviour> a = std::make_shared<_Ty>(aaa);
103 
104 
105 
106             CORE::Behaviour::behaviours\&.push_back(a);
107             CORE::Behaviour::AWAKES\&.push_back(a);
108             CORE::Behaviour::Starts\&.push_back(a);
109             behaviours\&.push_back(a);
110             TracyCZoneEnd(ctxx);
111 
112             return std::dynamic_pointer_cast<_Ty>(a)\&.get();
113         }
.PP
.fi

.PP
References \fBCORE::Behaviour::AWAKES\fP, \fBCORE::Behaviour::behaviours\fP, \fBbehaviours\fP, \fBname\fP, \fBCORE::Behaviour::Starts\fP, \fBTrAnSfOrM\fP, \fBCORE::Object::uuid\fP, and \fBuuids\fP\&.
.SS "template<class _Ty , class\&.\&.\&. _Types> void GameObject::AddComponent (_Ty arg, void * component)\fR [inline]\fP"

.PP
Definition at line \fB118\fP of file \fBGAMEOBJECT\&.h\fP\&.
.nf
118                                                     {
119             TracyCZoneN(ctxx, "Adding Component", true);
120 
121             if (arg\&.UUID == GUID_NULL) {
122                 UUID uuid;
123                 UuidCreate(&uuid);
124 
125 
126                 arg\&.UUID = uuid;
127                 arg\&.uuID = arg\&.UUID;
128             }
129             uuids\&.push_back(arg\&.UUID);
130             auto aa = (_Ty*)component;
131             aa\->gameobject = this;
132             //using name = decltype(_Ty);
133              std::string nm= typeid(_Ty)\&.name();
134              aa\->compname = nm;
135             //aa\->gameobject = this;
136             std::shared_ptr<CORE::Behaviour> a = std::make_shared<_Ty>(*aa);
137 
138 
139 
140             CORE::Behaviour::behaviours\&.push_back(a);
141             CORE::Behaviour::AWAKES\&.push_back(a);
142             CORE::Behaviour::Starts\&.push_back(a);
143             behaviours\&.push_back(a);
144 
145             Components\&.push_back(component);
146             TracyCZoneEnd(ctxx);
147 
148         }
.PP
.fi

.PP
References \fBCORE::Behaviour::AWAKES\fP, \fBCORE::Behaviour::behaviours\fP, \fBbehaviours\fP, \fBComponents\fP, \fBname\fP, \fBCORE::Behaviour::Starts\fP, \fBCORE::Object::uuid\fP, and \fBuuids\fP\&.
.SS "std::shared_ptr< \fBCORE::Behaviour\fP > GameObject::AddComponent (std::shared_ptr< \fBCORE::Behaviour\fP > component, std::string name)"

.PP
Definition at line \fB107\fP of file \fBGAMEOBJECT\&.cpp\fP\&.
.nf
107                                                                                                              {
108 
109     auto tmp = component\->Copy();
110     if (tmp\->uuID == GUID_NULL) {
111         UUID uuid;
112         UuidCreate(&uuid);
113 
114 
115         //tmp\->UUID = uuid;
116         tmp\->uuID = uuid;
117     }
118     tmp\->gameobject = this;
119     tmp\->TRANSFORM = this\->TrAnSfOrM;
120     tmp\->compname = name;
121     uuids\&.push_back(tmp\->uuID);
122     CORE::Behaviour::behaviours\&.push_back(tmp);
123     CORE::Behaviour::AWAKES\&.push_back(tmp);
124     CORE::Behaviour::Starts\&.push_back(tmp);
125     behaviours\&.push_back(tmp);
126     return tmp;
127 }
.PP
.fi

.PP
References \fBCORE::Behaviour::AWAKES\fP, \fBCORE::Behaviour::behaviours\fP, \fBbehaviours\fP, \fBname\fP, \fBCORE::Behaviour::Starts\fP, \fBTrAnSfOrM\fP, \fBCORE::Object::uuid\fP, and \fBuuids\fP\&.
.PP
Referenced by \fBCreate()\fP, \fBCreateCamera()\fP, and \fBCreateEmpty()\fP\&.
.SS "shared_ptr< \fBGameObject\fP > GameObject::Create (\fBDATATYPES::TS_P_Vector3\fP pos, \fBQuaternion\fP rot, std::shared_ptr< \fBCOMPONENTS::_Mesh\fP > mesh, int materialID)\fR [static]\fP"

.PP
Creates new \fBGameObject\fP\&. 
.PP
\fBParameters\fP
.RS 4
\fIpos\fP 
.br
\fImesh\fP 
.br
\fImaterialID\fP 
.RE
.PP
\fBReturns\fP
.RS 4
'shared_ptr GameObject'
.RE
.PP

.PP
Definition at line \fB47\fP of file \fBGAMEOBJECT\&.cpp\fP\&.
.nf
47                                                                                                                                        {
48     TracyCZoneN(ctx, "Creating GameObject", true);
49 
50     GameObject tmp;
51     UuidCreate(&tmp\&.uuid);
52     Transform transf = Transform();
53     transf\&.Position = pos;
54     transf\&.rotation = rot;
55     tmp\&.TrAnSfOrM = std::make_shared<Transform>(transf);
56     tmp\&.MESH = mesh;
57     tmp\&.TrAnSfOrM\->UpdateDirections();
58     tmp\&.MESH\->MMLnik = std::make_shared<PAIN::MiniModel>(PAIN::MiniModel(tmp\&.MESH\->Model, tmp\&.TrAnSfOrM, std::make_shared<bool>(tmp\&.Enabled), std::make_shared<bool>(tmp\&.isRemoved)));
59 
60     auto a = std::make_shared<GameObject>(tmp);
61     PAIN::Render::mats[materialID]\&.objects\&.push_back(a);
62     if (materialID < PAIN::Render::mats\&.size() || PAIN::Render::mats\&.size() == 0) {
63         PAIN::Render::mats[materialID]\&.Object_ModelSubstitute\&.push_back(tmp\&.MESH\->MMLnik);
64 
65     }
66     CORE::Scene::activeScene\->ObjectsInScene\&.push_back(a);
67     TracyCZoneEnd(ctx);
68 
69     return a;
70 
71 }
.PP
.fi

.PP
References \fBCORE::Scene::activeScene\fP, \fBEnabled\fP, \fBisRemoved\fP, \fBPAIN::Render::mats\fP, \fBMESH\fP, \fBTrAnSfOrM\fP, and \fBCORE::Object::uuid\fP\&.
.SS "shared_ptr< \fBGameObject\fP > GameObject::Create (\fBDATATYPES::TS_P_Vector3\fP pos, void * mesh, int materialID)\fR [static]\fP"

.PP
Definition at line \fB12\fP of file \fBGAMEOBJECT\&.cpp\fP\&.
.nf
12                                                                                               {
13     TracyCZoneN(ctx, "Creating GameObject", true);
14 
15     GameObject tmp;
16     UuidCreate(&tmp\&.uuid);
17     Transform transf = Transform();
18     transf\&.Position = pos;
19     void* tra = &transf;
20     tmp\&.msh = mesh;
21     tmp\&.AddComponent(COMPONENTS::_Mesh(), mesh);
22     //auto a = tmp\&.GetComponent(COMPONENTS::_Mesh());
23     tmp\&.MESH = std::make_shared<COMPONENTS::_Mesh>(*(COMPONENTS::_Mesh*)tmp\&.Components[0]);
24     //tmp\&.MESH = tmp\&.MesH;
25     tmp\&.AddComponent(Transform(), tra);
26     tmp\&.TrAnSfOrM = std::dynamic_pointer_cast<Transform>(tmp\&.behaviours[1]);
27     tmp\&.MESH\->MMLnik = std::make_shared<PAIN::MiniModel>(PAIN::MiniModel(tmp\&.MESH\->Model, tmp\&.TrAnSfOrM, std::make_shared<bool>(tmp\&.Enabled), std::make_shared<bool>(tmp\&.isRemoved)));
28 
29     auto a = std::make_shared<GameObject>(tmp);
30     if (materialID < PAIN::Render::mats\&.size() || PAIN::Render::mats\&.size() == 0) {
31         PAIN::Render::mats[materialID]\&.Object_ModelSubstitute\&.push_back(tmp\&.MESH\->MMLnik);
32     }
33     CORE::Scene::activeScene\->ObjectsInScene\&.push_back(a);
34     std::dynamic_pointer_cast<Transform>(tmp\&.behaviours[1])\&.get()\->UpdateDirections();
35     TracyCZoneEnd(ctx);
36 
37     return a;
38 
39 }
.PP
.fi

.PP
References \fBCORE::Scene::activeScene\fP, \fBAddComponent()\fP, \fBbehaviours\fP, \fBComponents\fP, \fBEnabled\fP, \fBisRemoved\fP, \fBPAIN::Render::mats\fP, \fBMESH\fP, \fBmsh\fP, \fBTrAnSfOrM\fP, and \fBCORE::Object::uuid\fP\&.
.PP
Referenced by \fBbb()\fP, and \fBprocessInput()\fP\&.
.SS "\fBGameObject\fP GameObject::CreateCamera (\fBDATATYPES::TS_P_Vector3\fP pos, \fBQuaternion\fP rot)\fR [static]\fP"

.PP
Definition at line \fB87\fP of file \fBGAMEOBJECT\&.cpp\fP\&.
.nf
88 {
89     TracyCZoneN(ctx, "Creating Camera", true);
90 
91     GameObject tmp;
92     UuidCreate(&tmp\&.uuid);
93     Transform transf = Transform();
94     transf\&.Position = pos;
95     transf\&.rotation = rot;
96     tmp\&.TrAnSfOrM = std::make_shared<Transform>(transf);
97     //Empty empt = Empty();
98     tmp\&.AddComponent(Empty());
99     
100 
101     tmp\&.AddComponent(Camera(tmp\&.TrAnSfOrM));
102     tmp\&.TrAnSfOrM\->UpdateDirections();
103     TracyCZoneEnd(ctx);
104 
105     return tmp;
106 }
.PP
.fi

.PP
References \fBAddComponent()\fP, \fBTrAnSfOrM\fP, and \fBCORE::Object::uuid\fP\&.
.PP
Referenced by \fBPAIN::RenderStup::Setup()\fP\&.
.SS "\fBGameObject\fP GameObject::CreateEmpty (\fBDATATYPES::TS_P_Vector3\fP pos)\fR [static]\fP"

.PP
Definition at line \fB72\fP of file \fBGAMEOBJECT\&.cpp\fP\&.
.nf
73 {
74     GameObject tmp;
75     UuidCreate(&tmp\&.uuid);
76     Transform transf = Transform();
77     transf\&.Position = pos;
78     void* tra = &transf;
79     Empty empt = Empty();
80     tmp\&.AddComponent(Empty(), &empt);
81     tmp\&.AddComponent(Transform(), tra);
82     tmp\&.TrAnSfOrM = std::dynamic_pointer_cast<Transform>(tmp\&.behaviours[1]);
83     std::dynamic_pointer_cast<Transform>(tmp\&.behaviours[1])\&.get()\->UpdateDirections();
84 
85     return tmp;
86 }
.PP
.fi

.PP
References \fBAddComponent()\fP, \fBbehaviours\fP, \fBTrAnSfOrM\fP, and \fBCORE::Object::uuid\fP\&.
.PP
Referenced by \fBCORE::ConfigLoader::GameobjectFromCFGFile()\fP\&.
.SS "template<class _Ty , class\&.\&.\&. _Types> _Ty * GameObject::GetComponent (_Ty arg)\fR [inline]\fP"

.PP
Definition at line \fB154\fP of file \fBGAMEOBJECT\&.h\fP\&.
.nf
154                                    {
155             TracyCZoneN(ctxx, "Getting  Component", true);
156 
157 
158             for (int i = 0; i < Components\&.size(); i++) {
159                 if (arg\&.UUID == uuids[i]) {
160 
161                     std::cout << "found";
162                     TracyCZoneEnd(ctxx);
163 
164                     return (_Ty*)Components[i];
165                 }
166 
167 
168             }
169             TracyCZoneEnd(ctxx);
170 
171         }
.PP
.fi

.PP
References \fBComponents\fP, \fBi\fP, and \fBuuids\fP\&.
.SS "template<class _Ty , class\&.\&.\&. _Types> _Ty * GameObject::GetComponentDynamic (_Ty arg)\fR [inline]\fP"

.PP
Definition at line \fB173\fP of file \fBGAMEOBJECT\&.h\fP\&.
.nf
173                                           {
174             TracyCZoneN(ctxx, "Getting  Component", true);
175 
176 
177             for (int i = 0; i < Components\&.size(); i++) {
178                 if (arg\&.UUID == uuids[i]) {
179 
180                     std::cout << "found";
181                     TracyCZoneEnd(ctxx);
182 
183                     return std::dynamic_pointer_cast<_Ty>(behaviours[i])\&.get();
184                 }
185 
186 
187             }
188             TracyCZoneEnd(ctxx);
189 
190         }
.PP
.fi

.PP
References \fBbehaviours\fP, \fBComponents\fP, \fBi\fP, and \fBuuids\fP\&.
.SS "\fBGameObject\fP & GameObject::operator= (const \fBGameObject\fP & other)\fR [inline]\fP"

.PP
Definition at line \fB55\fP of file \fBGAMEOBJECT\&.h\fP\&.
.nf
55                                                        {
56             if (this == &other) return *this; // self\-assignment guard
57             Enabled = other\&.Enabled;
58             MESH = other\&.MESH;
59             TrAnSfOrM = other\&.TrAnSfOrM;
60             MesH = other\&.MesH;
61             msh = other\&.msh;
62             name = other\&.name;
63             Components = other\&.Components;
64             behaviours = other\&.behaviours;
65             uuids = other\&.uuids;
66             return *this;
67         }
.PP
.fi

.PP
References \fBbehaviours\fP, \fBComponents\fP, \fBEnabled\fP, \fBMESH\fP, \fBMesH\fP, \fBmsh\fP, \fBname\fP, \fBTrAnSfOrM\fP, and \fBuuids\fP\&.
.SH "Member Data Documentation"
.PP 
.SS "std::vector<std::shared_ptr<\fBCORE::Behaviour\fP> > GameObject::behaviours"

.PP
Definition at line \fB37\fP of file \fBGAMEOBJECT\&.h\fP\&.
.PP
Referenced by \fBAddComponent()\fP, \fBAddComponent()\fP, \fBAddComponent()\fP, \fBCreate()\fP, \fBCreateEmpty()\fP, \fBGetComponentDynamic()\fP, and \fBoperator=()\fP\&.
.SS "std::vector<void*> GameObject::Components"

.PP
Definition at line \fB36\fP of file \fBGAMEOBJECT\&.h\fP\&.
.PP
Referenced by \fBAddComponent()\fP, \fBCreate()\fP, \fBGetComponent()\fP, \fBGetComponentDynamic()\fP, and \fBoperator=()\fP\&.
.SS "bool GameObject::Enabled = true"

.PP
Definition at line \fB26\fP of file \fBGAMEOBJECT\&.h\fP\&.
.PP
Referenced by \fBCreate()\fP, \fBCreate()\fP, and \fBoperator=()\fP\&.
.SS "const bool GameObject::isNull"

.PP
Definition at line \fB33\fP of file \fBGAMEOBJECT\&.h\fP\&.
.SS "bool GameObject::isRemoved"

.PP
Definition at line \fB34\fP of file \fBGAMEOBJECT\&.h\fP\&.
.PP
Referenced by \fBCreate()\fP, and \fBCreate()\fP\&.
.SS "\fBLayer\fP GameObject::layer"

.PP
Definition at line \fB31\fP of file \fBGAMEOBJECT\&.h\fP\&.
.SS "std::shared_ptr<\fBCOMPONENTS::_Mesh\fP> GameObject::MESH"

.PP
Definition at line \fB27\fP of file \fBGAMEOBJECT\&.h\fP\&.
.PP
Referenced by \fBCreate()\fP, \fBCreate()\fP, and \fBoperator=()\fP\&.
.SS "std::shared_ptr<\fBCORE::Behaviour\fP> GameObject::MesH"

.PP
Definition at line \fB29\fP of file \fBGAMEOBJECT\&.h\fP\&.
.PP
Referenced by \fBoperator=()\fP\&.
.SS "void* GameObject::msh"

.PP
Definition at line \fB32\fP of file \fBGAMEOBJECT\&.h\fP\&.
.PP
Referenced by \fBCreate()\fP, and \fBoperator=()\fP\&.
.SS "std::string GameObject::name"

.PP
Definition at line \fB35\fP of file \fBGAMEOBJECT\&.h\fP\&.
.PP
Referenced by \fBAddComponent()\fP, \fBAddComponent()\fP, \fBAddComponent()\fP, and \fBoperator=()\fP\&.
.SS "\fBTag\fP GameObject::tag"

.PP
Definition at line \fB30\fP of file \fBGAMEOBJECT\&.h\fP\&.
.SS "std::shared_ptr<\fBTransform\fP> GameObject::TrAnSfOrM"

.PP
Definition at line \fB28\fP of file \fBGAMEOBJECT\&.h\fP\&.
.PP
Referenced by \fBAddComponent()\fP, \fBAddComponent()\fP, \fBCreate()\fP, \fBCreate()\fP, \fBCreateCamera()\fP, \fBCreateEmpty()\fP, \fBPHYSICS::Physics::intersect()\fP, \fBoperator=()\fP, \fBTest::Update()\fP, and \fBPHYSICS::Physics::updateBox()\fP\&.
.SS "std::vector<UUID> GameObject::uuids"

.PP
Definition at line \fB38\fP of file \fBGAMEOBJECT\&.h\fP\&.
.PP
Referenced by \fBAddComponent()\fP, \fBAddComponent()\fP, \fBAddComponent()\fP, \fBGetComponent()\fP, \fBGetComponentDynamic()\fP, and \fBoperator=()\fP\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for My Project from the source code\&.


UPRISE_GAME/src/ECS/ENTITYS/GAMEOBJECT.cpp
#pragma once
#include "pch.h"
#include "Header/CORE/CORE.h"
#include "GAMEOBJECT.h"
#include "_COMPONENT.h"
#include "CAMERA.h"
#include "RENDER_MATERIAL.h"
#include "Empty.h"
#include "MESH.h"
#include "Header/CORE/C_SCENE.h"
#include "DLL-ENGINE-LINK.h"
//#include "CUSTOM_STD.h"

::shared_ptr<GameObject> GameObject::Create(DATATYPES::TSPVector3 pos, void* mesh, int materialID) {
	TracyCZoneN(ctx, "Creating GameObject", true);
	GOP;
	GameObject tmp;
	UuidCreate(&tmp.uuid);
	Transform transf = Transform();
	transf.Position = pos;
	void* tra = &transf;
	tmp.msh = mesh;
	tmp.AddComponent(COMPONENTS::_Mesh(), mesh);
	//auto a = tmp.GetComponent(COMPONENTS::_Mesh());
	tmp.MESH = std::make_shared<COMPONENTS::_Mesh>(*(COMPONENTS::_Mesh*)tmp.Components[0]);
	tmp.MESH->Material_ID = materialID;

	//tmp.MESH = tmp.MesH;
	tmp.AddComponent(Transform(), tra);
	tmp.TrAnSfOrM = std::dynamic_pointer_cast<Transform>(tmp.behaviours[1]);
	auto a = std::make_shared<GameObject>(tmp);
	CORE::Scene::activeScene->ObjectsInScene.push_back(a);
	std::dynamic_pointer_cast<Transform>(tmp.behaviours[1]).get()->UpdateDirections();
	TracyCZoneEnd(ctx);

	return a;

}
/// <summary>
/// Creates new GameObject 
/// </summary>
/// <param name="pos"></param>
/// <param name="mesh"></param>
/// <param name="materialID"></param>
/// <returns> "shared_ptr GameObject"</returns>
shared_ptr<GameObject> GameObject::Create(DATATYPES::TSPVector3 pos, Quaternion rot, std::shared_ptr<COMPONENTS::_Mesh> mesh, int materialID) {
	TracyCZoneN(ctx, "Creating GameObject", true);
	GOP;
	GameObject tmp;
	UuidCreate(&tmp.uuid);
	Transform transf = Transform();
	transf.Position = pos;
	transf.rotation = rot;
	tmp.TrAnSfOrM = std::make_shared<Transform>(transf);
	tmp.MESH = mesh;
	tmp.MESH->Material_ID = materialID;
	tmp.TrAnSfOrM->UpdateDirections();
	tmp.MESH->MMLnik = std::make_shared<PAIN::MiniModel>(PAIN::MiniModel(tmp.MESH->Model, tmp.TrAnSfOrM, std::make_shared<bool>(tmp.Enabled), std::make_shared<bool>(tmp.isRemoved)));
	auto a = std::make_shared<GameObject>(tmp);
	//auto enb = std::make_shared<bool>(a->Enabled);
	
	IMPORTANT::DATA->AddTORender((tmp.MESH->MMLnik), materialID);
	CORE::Scene::activeScene->ObjectsInScene.push_back(a);
	TracyCZoneEnd(ctx);

	return a;

}
GameObject GameObject::CreateEmpty(DATATYPES::TSPVector3 pos)
{
	GOP;
	GameObject tmp;
	UuidCreate(&tmp.uuid);
	Transform transf = Transform();
	transf.Position = pos;
	void* tra = &transf;
	Empty empt = Empty();
	tmp.AddComponent(Empty(), &empt);
	tmp.AddComponent(Transform(), tra);
	tmp.TrAnSfOrM = std::dynamic_pointer_cast<Transform>(tmp.behaviours[1]);
	std::dynamic_pointer_cast<Transform>(tmp.behaviours[1]).get()->UpdateDirections();

	return tmp;
}
GameObject GameObject::CreateCamera(DATATYPES::TSPVector3 pos, Quaternion rot)
{
	TracyCZoneN(ctx, "Creating Camera", true);
	GOP;
	GameObject tmp;
	tmp.IsCamera = true;
	UuidCreate(&tmp.uuid);
	Transform transf = Transform();
	transf.Position = pos;
	transf.rotation = rot;
	tmp.TrAnSfOrM = std::make_shared<Transform>(transf);
	//Empty empt = Empty();
	tmp.AddComponent(Empty());


	tmp.AddComponent(Camera(tmp.TrAnSfOrM));
	tmp.TrAnSfOrM->UpdateDirections();
	TracyCZoneEnd(ctx);

	return tmp;
}
bool GameObject::Delete(std::shared_ptr<GameObject> Object)
{
	if (std::find(CORE::Scene::Backups_SCENE->ObjectsInScene.begin(), CORE::Scene::Backups_SCENE->ObjectsInScene.end(), Object) != CORE::Scene::Backups_SCENE->ObjectsInScene.end()){
		int cnt = Object->behaviours.size() - 1;
		IMPORTANT::DATA->RemoveFromRender(Object->MESH->MMLnik, Object->MESH->Material_ID);
		GOM;
		for (; cnt >= 0; cnt--) {
			RemoveComponent(Object->behaviours[cnt]);
		}
		*Object->MESH->MMLnik->_removed = true;
		//GameObject* Replacer = new GameObject(1);
		Object->isRemoved = true;
		auto a = Object.get();
		//delete a;
		Object.DESTROY();
		 Object.setPTR(nullptr);
		//auto a = std::make_shared<GameObject>(Replacer);
		//==//=;

		//Replacer->isRemoved = true;
		//auto aa = Object.get();
		//*aa = *Replacer;
	}


	return true;
}
bool GameObject::Delete(GameObject& Object)
{
	GameObject Replacer(1);
	Object.isRemoved = true;
	//IMPORTANT::DATA->RemoveFromRender();
	//auto a = std::make_shared<GameObject>(Replacer);
	Object = Replacer;
	return false;
}
bool GameObject::Delete(GameObject* Object)
{
	GameObject Replacer(1);
	Object->isRemoved = true;
	//IMPORTANT::DATA->RemoveFromRender();
	//auto a = std::make_shared<GameObject>(Replacer);
	Object = &Replacer;

	return false;
}
bool GameObject::RemoveComponent(std::shared_ptr<CORE::Behaviour> Component)
{

	Empty replacer;
	replacer.ToBeRemoved = true;
	std::shared_ptr<CORE::Behaviour> a = std::make_shared<Empty>(replacer);
	Component.swap(a);
	return false;
}
std::shared_ptr<CORE::Behaviour> GameObject::AddComponent(std::shared_ptr<CORE::Behaviour> component, std::string name) {

	auto tmp = component->Copy();
	if (tmp->uuID == GUID_NULL) {
		UUID uuid;
		UuidCreate(&uuid);


		//tmp->UUID = uuid;
		tmp->uuID = uuid;
	}
	tmp->Game_Object = this;
	tmp->TRANSFORM = this->TrAnSfOrM;
	tmp->compname = name;
	uuids.push_back(tmp->uuID);
	CORE::Behaviour::behaviours.push_back(tmp);
	CORE::Behaviour::AWAKES.push_back(tmp);
	CORE::Behaviour::Starts.push_back(tmp);
	behaviours.push_back(tmp);
	return tmp;
}
//void GameObject::AddComponent(void* component) {
//
//	Components.push_back(component);
//
//}
//template <class _Ty, class... _Types>

//void GameObject::AddComponent(_Ty arg, void* component) {
//	if (arg.UUID == GUID_NULL) {
//		UUID uuid;
//		UuidCreate(&uuid);
//
//
//		arg.UUID = uuid;
//		arg.uuID = arg.UUID;
//	}
//	uuids.push_back(arg.UUID);
//	auto aa = (_Ty*)component;
//	//aa->gameobject = this;
//	std::shared_ptr<CORE::Behaviour> a = std::make_shared<_Ty>(*aa);
//
//
//
//	CORE::Behaviour::behaviours.push_back(a);
//	behaviours.push_back(a);
//
//	Components.push_back(component);
//}

//template<class _Ty, class ..._Types>
//void GameObject::AddComponent(_Ty arg, std::shared_ptr<CORE::Behaviour> component,void* voidcomp)
//{
//	if (arg.UUID == GUID_NULL) {
//		UUID uuid;
//		UuidCreate(&uuid);
//
//
//		arg.UUID = uuid;
//		arg.uuID = arg.UUID;
//	}
//	uuids.push_back(arg.UUID);
//	CORE::Behaviour::behaviours.push_back(a);
//	behaviours.push_back(a);
//	Components.push_back(component);
//
//
//
//}

//template <class _Ty, class... _Types>
//_Ty* GameObject::GetComponent(_Ty arg) {
//
//
//	for (int i = 0; i < Components.size(); i++) {
//		if (arg.UUID == uuids[i]) {
//
//			std::cout << "found";
//			return (_Ty*)Components[i];
//		}
//
//
//	}
//}


